### PR TITLE
[WIP] Complex dtype support

### PR DESCRIPTION
**Pull request by Randl**
Fix https://github.com/google/lineax/issues/57

- [ ] Add `dtype` parameter to tests, test with both real and complex number
  - [ ] make sure data types are explicit in all tests
- [x] Change typing from `Float` to `Inexact`
- [x] Add `holomorphic` to appropriate places in `jaxfwd`
- [ ] Add complex-related tags and methods, e.g., `adjoint` (as opposed to `transpose`)
- [ ] Fix complex-specific bugs in methods, e.g., missing conjugates, make sure tests are passing
- [ ] Add complex-specific and new functionality (such as adjoint) tests

### CODE REVIEW

**Comment by Randl** [comment]
If I put float64/complex128 in tests, most of them pass, except for a number of jvp tests wtih singular complex matrices for SVD and some transpose tests. However, with float32/complex64, additional tests fail (due to lack of precision?). So three main questions are which types to use in tests to make sure that everything works, which additional tests are needed specifically for complex variables, and what is the problem with SVD.
@patrick-kidger any insights?

**Comment by patrick-kidger** [comment]
Thanks for starting in on this! This is really cool to see.

Regarding the float32 tests, this is expected. We actually set float64 as the default for any test which doesn't already specify it explicitly: https://github.com/google/lineax/blob/9a573529f3eea2a6944009f6e4afe4f364366baf/tests/conftest.py#L24
It's fairly common for linear solves to be a bit finickity on this front.  
So likewise, I'd be happy to use just complex128 in the tests.

Regarding "which additional tests are needed specifically for complex variables". A couple of things come to mind:
- That `_misc.two_norm` and `_misc.max_norm` are adjusted to do the right thing, along with their JVPs. This should actually be fairly easy as I actually did this in Optimistix [here](https://github.com/patrick-kidger/optimistix/blob/4e97e487ce30506e34744bbb68423b7c5f330180/optimistix/_misc.py#L126-L198) with associated tests [here](https://github.com/patrick-kidger/optimistix/blob/4e97e487ce30506e34744bbb68423b7c5f330180/tests/test_misc.py#L39-L91). I think we just need to upstream the implementation to Lineax. (And side note, probably then make them available under `lineax.internal.{two,max}_norm` so we can just have Optimistix depend on the implementation here.)
- That we get transposes, not conjugate-transposes, when performing `jax.linear_transpose` of a `lineax.linear_solve`. (This is the convention JAX uses here for complex numbers.)

This is in addition to the tests for "getting the right forward solution" and "getting the right gradient" etc., which can be handled by extending existing tests.

---

You note that there's a few things that need adjusting. I'd ask that we do each piece across separate PRs, to make it easier for me to review them.

**Comment by Randl** [comment]
@patrick-kidger I've split the changes in 5 PRs, which I believe should be non-breaking if merged in order (the dtype one should be non-breaking by itself). 
Since the changes are not complete yet, these PRs do not fully cover the changes; in particular, the tests for complex dtype are still off.

**Comment by patrick-kidger** [comment]
Great, thank you -- I've gone through and left a few comments in some PRs. Overall I like the look of these changes!

**Comment by patrick-kidger** [comment]
Closing in favour of the individual PRs.
